src/main/java/net/snowflake/ingest/streaming/internal/IcebergParquetValueParser.java

```diff
@@ -298,4 +359,177 @@ private static int timeUnitToScale(LogicalTypeAnnotation.TimeUnit timeUnit) {
+          isDescendantsOfRepeatingGroup);
+    }
+    throw new SFException(
+        ErrorCode.UNKNOWN_DATA_TYPE, logicalTypeAnnotation, type.getClass().getSimpleName());
+  }
+
+  /**
+   * Parses a struct value based on Parquet group logical type. The parsed value is a list of
+   * values, where each element represents a field in the group. For example, an input {@code
+   * {"field1": 1, "field2": 2}} will be parsed as {@code [1, 2]}.
+   */
+  private static ParquetBufferValue getStructValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path,
+      boolean isDescendantsOfRepeatingGroup) {
+    Map<String, ?> structVal =
+        DataValidationUtil.validateAndParseIcebergStruct(path, value, insertRowsCurrIndex);
+    Set<String> extraFields = structVal.keySet();
+    List<Object> listVal = new ArrayList<>(type.getFieldCount());
+    float estimatedParquetSize = 0f;
+    for (int i = 0; i < type.getFieldCount(); i++) {
+      ParquetBufferValue parsedValue =
+          parseColumnValueToParquet(
+              structVal.getOrDefault(type.getFieldName(i), null),
+              type.getType(i),
+              statsMap,
+              defaultTimezone,
+              insertRowsCurrIndex,
+              path,
+              isDescendantsOfRepeatingGroup);
+      extraFields.remove(type.getFieldName(i));
+      listVal.add(parsedValue.getValue());
+      estimatedParquetSize += parsedValue.getSize();
+    }
+    if (!extraFields.isEmpty()) {
+      extraFields =
+          extraFields.stream().map(f -> concatDotPath(path, f)).collect(Collectors.toSet());
+      throw new SFException(
+          ErrorCode.INVALID_FORMAT_ROW,
+          "Extra fields: " + extraFields,
+          String.format(
+              "Fields not present in the struct shouldn't be specified, rowIndex:%d",
+              insertRowsCurrIndex));
+    }
+    return new ParquetBufferValue(listVal, estimatedParquetSize);
+  }
+
+  /**
+   * Parses an iterable value based on Parquet 3-level list logical type. Please check <a
+   * href="https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists">Parquet
+   * Logical Types#Lists</a> for more details. The parsed value is a list of lists, where each inner
+   * list represents a list of elements in the group. For example, an input {@code [1, 2, 3, 4]}
+   * will be parsed as {@code [[1], [2], [3], [4]]}.
+   */
+  private static ParquetBufferValue get3LevelListValue(
+      Object value,
+      GroupType type,
+      Map<String, RowBufferStats> statsMap,
+      ZoneId defaultTimezone,
+      final long insertRowsCurrIndex,
+      String path) {
+    Iterable<?> iterableVal =
+        DataValidationUtil.validateAndParseIcebergList(path, value, insertRowsCurrIndex);
+    List<Object> listVal = new ArrayList<>();
+    final AtomicReference<Float> estimatedParquetSize = new AtomicReference<>(0f);
```

Comment:
we don't need atomic here because there's no concurrency to manage. It introduces unnecessary volatile reads/writes.
if you do a regular for() loop you won't need the atomicRefrence and can live with regular float?